python/ray/workflow/tests/test_checkpoint.py

```diff
@@ -0,0 +1,236 @@
+import ray
+import time
+import pytest
+from ray.tests.conftest import *  # noqa
+
+import numpy as np
+from ray import workflow
+from ray.workflow.tests import utils
+from ray.workflow import storage, workflow_storage
+from ray.exceptions import RaySystemError
+
+
+@workflow.step
+def large_input():
+    return np.arange(2**24)
+
+
+@workflow.step
+def identity(x):
+    return x
+
+
+@workflow.step
+def average(x):
+    return np.mean(x)
+
+
+@workflow.step
+def checkpoint_dag(checkpoint):
+    x = large_input.options(name="large_input", checkpoint=checkpoint).step()
+    y = identity.options(name="identity", checkpoint=checkpoint).step(x)
+    return average.options(name="average").step(y)
+
+
+def _assert_step_checkpoints(wf_storage, step_id, mode):
+    result = wf_storage.inspect_step(step_id)
+    if mode == "all_skipped":
+        assert not result.output_object_valid
+        assert result.output_step_id is None
+        assert not result.args_valid
+        assert not result.func_body_valid
+        assert not result.step_options
+    elif mode == "output_skipped":
+        assert not result.output_object_valid
+        assert result.output_step_id is None
+        assert result.args_valid
+        assert result.func_body_valid
+        assert result.step_options is not None
+    elif mode == "checkpointed":
+        assert (result.output_object_valid
+                or result.output_step_id is not None)
+    else:
+        raise ValueError("Unknown mode.")
+
+
+def test_checkpoint_dag(workflow_start_regular):
+    # warm up
+    for _ in range(3):
+        outputs = checkpoint_dag.step(True).run()
+        assert np.isclose(outputs, 8388607.5)
+
+    global_storage = storage.get_global_storage()
+
+    # =================== skip all checkpoints ===================
+
+    start = time.time()
+    outputs = checkpoint_dag.options(
+        name="checkpoint_dag",
+        checkpoint=False).step(False).run(workflow_id="checkpoint_skip")
+    run_duration_skipped = time.time() - start
+    assert np.isclose(outputs, 8388607.5)
+
+    wf_storage = workflow_storage.WorkflowStorage("checkpoint_skip",
+                                                  global_storage)
+    _assert_step_checkpoints(
+        wf_storage, "checkpoint_dag", mode="output_skipped")
+    _assert_step_checkpoints(wf_storage, "large_input", mode="all_skipped")
+    _assert_step_checkpoints(wf_storage, "identity", mode="all_skipped")
+    _assert_step_checkpoints(wf_storage, "average", mode="all_skipped")
+
+    start = time.time()
+    recovered = ray.get(workflow.resume("checkpoint_skip"))
```

Comment:
Is there a way we can check whether `recovered` actually caused the checkpoint=False workflow step to re-execute? I think that's what we're actually interested in, not whether the checkpoint files exist or not (which is implementation-dependent).